File: runme/mcgen/local/datan/exasrc/e6lsq.f
      PROGRAM E6LSQ
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER (MAXN=20,MAXNR=2,N=20,NR=2)
      DIMENSION T(MAXN),Y(MAXN),DELTAY(MAXN),LIST(MAXNR),
     +X(MAXNR),CX(MAXNR,MAXNR),GX(MAXNR,MAXNR),A(MAXN,MAXNR),
     +DXPLUS(MAXNR),DXMINS(MAXNR),XPL(2),YPL(2)
      CHARACTER*75 STRING
      DOUBLE PRECISION LSQEXP
      EXTERNAL LSQEXP
C identify program to user
      WRITE(*,*)' Program E6LSQ demonstrates use of LSQASN'
      WRITE(*,*)' '
C create simulated data
      X(1)=10.D0
      X(2)=1.D0
      NSEED1=15
      NSEED2=211
      CALL RNE2IN(NSEED1,NSEED2)
      CALL RNSTNR(DELTAY,N)
      CALL RNECUY(Y,N)
      DELT=DBLE(2./21.)
      SIGMA=2.D0
      DO 10 I=1,N
        T(I)=DBLE(I)*DELT
        SIG=SIGMA*(DBLE(0.5)+Y(I))
        Y(I)=LSQEXP(X,NR,T(I))+DELTAY(I)*SIG
        DELTAY(I)=SIG
10    CONTINUE
C set first approximation of unknowns and perform fit with LSQNON
      NRED=NR
      X(1)=.1D0
      X(2)=.1D0
      LIST(1)=1
      LIST(2)=1
      NSTEP=100
      CALL LSQNON(LSQEXP,T,Y,DELTAY,N,NR,NRED,
     +LIST,X,CX,R,A,GX,NSTEP)
      NSTEP=100
C compute asymmetric errors
      CALL LSQASN(LSQEXP,T,Y,DELTAY,N,NR,NRED,LIST,X,CX,R,0.D0,
     +DXPLUS,DXMINS,A,GX,NSTEP)
C prepare graphics
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      CALL GROPEN
      CALL GROPWS(NWS)
      XMIN=5.D0
      YMIN=.5D0
      XMAX=15.D0
      YMAX=1.5D0
      CALL GRWNCC(XMIN,XMAX,YMIN,YMAX)
      CALL GRVWWC(0.2D0,0.9D0,0.2D0,0.9D0)
      CALL GRWNWC(-.414D0,1.D0,0.D0,1.D0)
      CALL GRFRAM
      CALL GRBOUN
      STRING='x_1'
      CALL GRSCLX(STRING,75)
      STRING='x_2'
      CALL GRSCLY(STRING,75)
      DX1=SQRT(CX(1,1))
      DX2=SQRT(CX(2,2))
      RHO=CX(1,2)/(DX1*DX2)
      CALL GRSTCL(2)
      CALL GRDATP(1,1.D0,X(1),X(2),DX1,DX2,RHO)
      CALL GRSTCL(3)
      DO 20 I=1,2
        IF(I.EQ.1) THEN
          XPL(1)=X(1)-DXMINS(1)
        ELSE
          XPL(1)=X(1)+DXPLUS(1)
        END IF
        XPL(2)=XPL(1)
        YPL(1)=YMIN
        YPL(2)=YMAX
        CALL GRPLIN(2,XPL,YPL)
20    CONTINUE
      DO 30 I=1,2
        IF(I.EQ.1) THEN
          YPL(1)=X(2)-DXMINS(2)
        ELSE
          YPL(1)=X(2)+DXPLUS(2)
        END IF
        YPL(2)=YPL(1)
        XPL(1)=XMIN
        XPL(2)=XMAX
        CALL GRPLIN(2,XPL,YPL)
30    CONTINUE
      FCONT=R+1.D0
      CALL GRSTCL(4)
      NX=100
      NY=100
      DX=(XMAX-XMIN)/NX
      DY=(YMAX-YMIN)/NY
C compute and draw contour with LSQCON
      CALL LSQCON(1,2,XMIN,YMIN,DX,DY,NX,NY,FCONT,
     +X,T,Y,DELTAY,N,NR,LSQEXP)
      CALL GRCLSE
      END

File: runme/mcgen/local/datan/exasrc/e6lsq.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__1 = 1, c__3 = 3, c__4 = 4, c__2 = 2, c__20 = 20, c__75 = 75;
static double c_d0 = 0., c_dp2 = .2, c_dp9 = .9, c_dmp414 = -.414, c_d1 = 1.;

main(void)
{
    /* Local variables */
    static integer i, nseed1, nseed2, nstep, nred, list[2], nx, ny, nws;
    static double xmin, ymin, xmax, ymax, a[40] /* was [20][2] */,
                  cx[4] /* was [2][2] */, dx, dy, gx[4] /* was [2][2] */,
                  r, t[20], x[2], y[20], sigma, fcont, deltay[20], dxmins[2],
                  dx1, dx2, dxplus[2], sig, rho, xpl[2], ypl[2], delt;
    static char string[75];

/* identify program to user */
    printf("%s\n\n", "Program E6LSQ demonstrates use of LSQASN.");
/* create simulated data */
    x[0] = 10.;
    x[1] = 1.;
    nseed1 = 15;
    nseed2 = 211;
    rne2in_(&nseed1, &nseed2);
    rnstnr_(deltay, &c__20);
    rnecuy_(y, &c__20);
    delt = .09523809523809523;
    sigma = 2.;
    for (i = 1; i <= 20; ++i) {
	t[i - 1] = (double) i * delt;
	sig = sigma * (.5 + y[i - 1]);
	y[i - 1] = lsqexp_(x, &c__2, &t[i - 1]) + deltay[i - 1] * sig;
	deltay[i - 1] = sig;
    }
/* set first approximation of unknowns and perform fit with LSQNON */
    nred = 2;
    x[0] = .1;
    x[1] = .1;
    list[0] = 1;
    list[1] = 1;
    nstep = 100;
    lsqnon_(lsqexp_, t, y, deltay, &c__20, &c__2, &nred, list, x, cx, &r, a,
           gx, &nstep);
    nstep = 100;
/* compute asymmetric errors */
    lsqasn_(lsqexp_, t, y, deltay, &c__20, &c__2, &nred, list, x, cx, &r,
           &c_d0, dxplus, dxmins, a, gx, &nstep);
/* prepare graphics */
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%ld", &nws);
    (void)getchar();
    gropen_();
    gropws_(&nws);
    xmin = 5.;
    ymin = .5;
    xmax = 15.;
    ymax = 1.5;
    grwncc_(&xmin, &xmax, &ymin, &ymax);
    grvwwc_(&c_dp2, &c_dp9, &c_dp2, &c_dp9);
    grwnwc_(&c_dmp414, &c_d1, &c_d0, &c_d1);
    grfram_();
    grboun_();
    memset(string, ' ', 75);
    strncpy(string, "x_1", 3);
    grsclx_(string, &c__75);
    strncpy(string, "x_2", 3);
    grscly_(string, &c__75);
    dx1 = sqrt(cx[0]);
    dx2 = sqrt(cx[3]);
    rho = cx[2] / (dx1 * dx2);
    grstcl_(&c__2);
    grdatp_(&c__1, &c_d1, x, &x[1], &dx1, &dx2, &rho);
    grstcl_(&c__3);
    for (i = 1; i <= 2; ++i) {
	if (i == 1) {
	    xpl[0] = x[0] - dxmins[0];
	} else {
	    xpl[0] = x[0] + dxplus[0];
	}
	xpl[1] = xpl[0];
	ypl[0] = ymin;
	ypl[1] = ymax;
	grplin_(&c__2, xpl, ypl);
    }
    for (i = 1; i <= 2; ++i) {
	if (i == 1) {
	    ypl[0] = x[1] - dxmins[1];
	} else {
	    ypl[0] = x[1] + dxplus[1];
	}
	ypl[1] = ypl[0];
	xpl[0] = xmin;
	xpl[1] = xmax;
	grplin_(&c__2, xpl, ypl);
    }
    fcont = r + 1.;
    grstcl_(&c__4);
    nx = 100;
    ny = 100;
    dx = (xmax - xmin) / nx;
    dy = (ymax - ymin) / ny;
/* compute and draw contour with LSQCON */
    lsqcon_(&c__1, &c__2, &xmin, &ymin, &dx, &dy, &nx, &ny, &fcont, x, t, y, 
	   deltay, &c__20, &c__2, lsqexp_);
    grclse_();
    return 0;
} /* main */

